sled-agent/src/artifact_store.rs

```diff
@@ -0,0 +1,916 @@
+            // 1. Open the mountpoint and its temp dir so we can fsync them at
+            // the end.
+            let parent_dir = match File::open(&mountpoint).await {
+                Ok(dir) => dir,
+                Err(err) => {
+                    log_and_store!(
+                        last_error, &self.log, "open", mountpoint, err
+                    );
+                    continue;
+                }
+            };
+            let temp_dir_path = mountpoint.join(TEMP_SUBDIR);
+            let temp_dir = match File::open(&temp_dir_path).await {
+                Ok(dir) => dir,
+                Err(err) => {
+                    log_and_store!(
+                        last_error,
+                        &self.log,
+                        "open",
+                        temp_dir_path,
+                        err
+                    );
+                    continue;
+                }
+            };
+            // 2. fsync the file.
+            if let Err(err) = file.as_file_mut().sync_all().await {
+                let path = file.path().to_owned();
+                log_and_store!(last_error, &self.log, "sync", path, err);
+                continue;
+            }
+            // 3. Rename temporary file.
+            let final_path = mountpoint.join(self.sha256.to_string());
+            let moved_final_path = final_path.clone();
+            if let Err(err) = tokio::task::spawn_blocking(move || {
+                file.persist(&moved_final_path)
+            })
+            .await?
+            {
+                error!(
+                    &self.log,
+                    "Failed to rename temporary file";
+                    "error" => &err.error,
+                    "from" => err.file.path().as_str(),
+                    "to" => final_path.as_str(),
+                );
+                last_error = Some(Error::FileRename {
+                    from: err.file.path().to_owned(),
+                    to: final_path,
+                    err: err.error,
+                });
+                continue;
+            }
+            // 4. fsync the parent directory for both the final path and its
+            // previous path.
+            if let Err(err) = parent_dir.sync_all().await {
+                log_and_store!(last_error, &self.log, "sync", mountpoint, err);
+                continue;
+            }
+            if let Err(err) = temp_dir.sync_all().await {
+                log_and_store!(
+                    last_error,
+                    &self.log,
+                    "sync",
+                    temp_dir_path,
+                    err
+                );
+                continue;
+            }
+
+            any_success = true;
+        }
+
+        if any_success {
+            info!(
+                &self.log,
+                "Wrote artifact";
+                "sha256" => &self.sha256.to_string(),
+            );
+            Ok(())
```

Comment:
Seems a little odd to me that this returns "Ok()" if any files were written. E.g., if we failed to write to both M.2s, but wrote a single file, this would return "success", which seems misleading IMO.

Do you think we should return an error, if **any** error is returned, and return "Ok()" at a higher level if **either** write is successfully propagated to one of the two M.2s?

Reply:
I think you're right; I'm going to verbosely type out the two cases where I intend this API to be used based on the design in my head so far.

1. When an operator uploads a repo, it's unpacked in Nexus-local storage and verified, then the repository is added to the database in a `replicating` status. That Nexus starts a saga to replicate all the artifacts across some minimum number of sleds (probably 3); once this is complete, the repository is set to `available`. If that particular Nexus dies and loses its storage, some other Nexus will need to fail the saga and set the repository to `failed`, and the operator will need to reupload the repository. (I am not sure if a saga is the right primitive here but we can talk about the larger design I have in chat somewhere.)
   
   During this saga **Nexus probably needs sled agent to return an error** in this partial write situation so that Nexus can pick another sled to try and write the artifacts to.
2. As a background task, Nexus will periodically ask all the sleds about all the artifacts they have, and compare that to the list of artifacts that they should have based on the database. It will then submit requests to sled agents to delete superfluous artifacts or copy missing artifacts from other sleds. In this situation, it doesn't matter whether the sled agent returns an error here or not; if we successfully write to only a single M.2, Nexus will eventually see that and submit another request to rectify that during its background task.